src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs

```diff
@@ -0,0 +1,299 @@
+                {
+                    _log.LogTrace("Completing link");
+                    link.Target.Complete();
+                }
+            }
+
+            _taskCompletionSource.TrySetResult(null);
+        }
+
+
+        private void CombineDeltas(List<ResourceEvent<TResource>> deltas)
+        {
+            if (deltas.Count < 2) return;
+            if (deltas.First().EventFlags.HasFlag(EventTypeFlags.Sync)) // if we had a sync item queued up and got something else, get rid of sync
+                deltas.RemoveAt(0);
+            if (deltas.Count < 2) return;
+
+            // if the entire object was created and removed before worker got a chance to touch it and worker has not chose to see these
+            // types of events, we can just get rid of this "transient" object and not even notify worker of its existence
+            if (_skipTransient && deltas[0].EventFlags.HasFlag(EventTypeFlags.Add) && deltas[deltas.Count - 1].EventFlags.HasFlag(EventTypeFlags.Delete))
+                deltas.Clear();
+        }
+
+        private void QueueActionLocked(ResourceEvent<TResource> obj)
+        {
+            var id = KeyOf(obj);
+            var exists = _items.TryGetValue(id, out var deltas);
+            if (!exists)
+            {
+                deltas = new List<ResourceEvent<TResource>>();
+                _items[id] = deltas;
+                _queue.Enqueue(deltas);
+                _log.LogTrace("Item not already know, adding to queue");
+            }
+
+            deltas.Add(obj);
+            CombineDeltas(deltas);
+        }
+
+        private void OfferMessagesToLinks()
+        {
+            _log.LogTrace("Offering messages to links");
+            List<TargetLink> linksToOfferTo;
+            lock (_lock)
+            {
+                linksToOfferTo = _targets.Where(x => x.LastOfferedMessageReply != DataflowMessageStatus.Postponed).ToList();
+            }
+
+            foreach (var link in linksToOfferTo)
+            {
+                DataflowMessageStatus reply;
+                do // keep feeding the link messages until queue is either empty or it tells us that it can't handle any more
+                {
+                    reply = OfferMessageToLink(link);
+                } while (reply == DataflowMessageStatus.Accepted);
+            }
+        }
+
+        private DataflowMessageStatus OfferMessageToLink(TargetLink link)
+        {
+            List<ResourceEvent<TResource>> msg;
+            lock (_lock)
+            {
+                if (!_queue.TryPeek(out msg))
+                {
+                    _log.LogTrace("Nothing to offer - queue is empty");
+                    return DataflowMessageStatus.NotAvailable; // queue is empty
+                }
+            }
+
+            var header = new DataflowMessageHeader(Interlocked.Increment(ref _msgId));
+            link.LastOfferedMessageReply = link.Target.OfferMessage(header, msg, this, true);
```

Comment:
Is it safe to mutate `link` without the lock held? This method can be invoked concurrently.